packages/build-utils/src/fs/run-user-scripts.ts

```diff
@@ -319,55 +329,84 @@ export async function walkParentDirs({
-  const opts: SpawnOptionsExtended = { cwd: destPath, ...spawnOpts };
-  const env = opts.env ? { ...opts.env } : { ...process.env };
-  delete env.NODE_ENV;
-  opts.env = getEnvForPackageManager({
-    cliType,
-    lockfileVersion,
-    nodeVersion,
-    env,
-  });
-  let commandArgs: string[];
+  try {
+    await runNpmInstallSema.acquire();
+    const { cliType, packageJsonPath, lockfileVersion } = await scanParentDirs(
+      destPath
+    );
+
+    // Only allow `runNpmInstall()` to run once per `package.json`
+    // when doing a default install (no additional args)
+    if (meta && packageJsonPath && args.length === 0) {
+      if (typeof meta.runNpmInstallSet === 'undefined') {
+        meta.runNpmInstallSet = new Set<string>();
+      }
+      if (isSet<string>(meta.runNpmInstallSet)) {
+        if (meta.runNpmInstallSet.has(packageJsonPath)) {
+          return false;
+        } else {
+          meta.runNpmInstallSet.add(packageJsonPath);
+        }
+      }
+    }
 
-  if (cliType === 'npm') {
-    opts.prettyCommand = 'npm install';
-    commandArgs = args
-      .filter(a => a !== '--prefer-offline')
-      .concat(['install', '--no-audit', '--unsafe-perm']);
-  } else if (cliType === 'pnpm') {
-    // PNPM's install command is similar to NPM's but without the audit nonsense
-    // @see options https://pnpm.io/cli/install
-    opts.prettyCommand = 'pnpm install';
-    commandArgs = args
-      .filter(a => a !== '--prefer-offline')
-      .concat(['install', '--unsafe-perm']);
-  } else {
-    opts.prettyCommand = 'yarn install';
-    commandArgs = ['install', ...args];
-  }
+    debug(`Installing to ${destPath}`);
 
-  if (process.env.NPM_ONLY_PRODUCTION) {
-    commandArgs.push('--production');
-  }
+    const opts: SpawnOptionsExtended = { cwd: destPath, ...spawnOpts };
+    const env = opts.env ? { ...opts.env } : { ...process.env };
+    delete env.NODE_ENV;
+    opts.env = getEnvForPackageManager({
+      cliType,
+      lockfileVersion,
+      nodeVersion,
+      env,
+    });
+    let commandArgs: string[];
+
+    if (cliType === 'npm') {
+      opts.prettyCommand = 'npm install';
+      commandArgs = args
+        .filter(a => a !== '--prefer-offline')
+        .concat(['install', '--no-audit', '--unsafe-perm']);
+    } else if (cliType === 'pnpm') {
+      // PNPM's install command is similar to NPM's but without the audit nonsense
+      // @see options https://pnpm.io/cli/install
+      opts.prettyCommand = 'pnpm install';
+      commandArgs = args
+        .filter(a => a !== '--prefer-offline')
+        .concat(['install', '--unsafe-perm']);
+    } else {
+      opts.prettyCommand = 'yarn install';
+      commandArgs = ['install', ...args];
+    }
 
-  return spawnAsync(cliType, commandArgs, opts);
+    if (process.env.NPM_ONLY_PRODUCTION) {
+      commandArgs.push('--production');
+    }
+
+    await spawnAsync(cliType, commandArgs, opts);
+    return true;
```

Comment:
I didn't realize that `finally` blocks would still run after `return` statements are executed. Huh.